### DISCUSSION TITLE
[2.6] deep model data binding - Indirect modification of overloaded element of Illuminate\Database\Eloquent\Collection has no effect

### DISCUSSION THREAD

**Discussion by takeshiyu** [livewire/livewire#3798] [1. Help] [answered]
Based on https://laravel-news.com/livewire-v2-6-0, Livewire has supported deep model data binding. And I write the following codes to test the feature:

```php
$details = new Collection([
    new Detail(['quantity' => 10, 'amount' => 5]),
]);

$this->receive = new Receive;
$this->receive->setRelation('details', $details);
```

```html
<div>
    @foreach($receive->details as $index => $detail)
    <input type="text" wire:model="receive.details.{{ $index }}.quantity" wire:key="receive.details.{{ $index }}.quantity">
    @endforeach
</div>
```

When I changed the `quantity`, I got the following error:

```
Indirect modification of overloaded element of Illuminate\Database\Eloquent\Collection has no effect
```

But, if I query the `receive` from database, it works fine.

```php
$this->receive = Receive::find(1);
```

Any idea what could be happening here? Any help or guidance is much appreciated.

**Comment by joshhanley** [comment]
@s12i hmm there was some issues with the deep nesting that we patched yesterday in 2.6.4.

Can you upgrade to that and test it again and let us know if you are still having issues?

Thanks!

**Reply by takeshiyu** [reply]
@joshhanley I try the `2.6.5`, and I still get the same error:

```
Indirect modification of overloaded element of Illuminate\Database\Eloquent\Collection has no effect {"userId":1,"exception":"[object] (ErrorException(code: 0): 
Indirect modification of overloaded element of Illuminate\\Database\\Eloquent\\Collection has no effect at /Users/s12i/repos/projects/hubs/vendor/laravel/framework/src/Illuminate/Collections/helpers.php:121)
```

My Laravel version is `8.61.0`

**Reply by joshhanley** [reply]
@s12i ok, no worries, thanks for testing! I will have a dig and see if I can work out what is going on.

Thanks!

**Comment by joshhanley** [comment, marked as answer]
@s12i ok so I've investigated this, and the issue is the fact that the detail model doesn't exist, so what happens is when it is serialised and then deserialised, Livewire has no way of knowing that an empty detail model should be created.

So before it's serialised it looks like this

```php
$details = new Collection([
  new Detail([
    'quantity' => 10,
    'amount' => 5
  ])
]);
```

But after it is serialised and then restored, it looks like this

```php
$details = new Collection([
  [
    'quantity' => 10,
    'amount' => 5
  ]
]);
```

Hence why the collection is throwing an overloaded error as there is no longer a detail model.

You can test this yourself by using the first `$details` object above and run this with it

```php
$key = '0.quantity';

$value = 20;

data_set($details, $key, $value);
```

You will see the first one runs happily, because there is a detail model already there.

Then re run the same code but this time with the second `$details` object.

This time you will see the overloaded element error.

I think that is why it's best to create and persist details models before you attach them as a relation to your `receive` model.

Hope this helps!

**Reply by takeshiyu** [reply]
@joshhanley  
Thank you for the help, and figure out the problem!
 It seems to me that I should use [Wireable objects](https://laravel-news.com/livewire-v2-6-0) when I want to create a `new` model before persistence, right ?

Thanks again for your support.

**Reply by joshhanley** [reply]
@s12i no worries 🙂 yeah wireable objects will work or stand alone models assigned directly to a property will also work, it's just the deep nesting with new models that doesn't. Or even just arrays.

Hope this helps!
